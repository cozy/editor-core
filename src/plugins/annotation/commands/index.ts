import type { EditorState } from '@atlaskit/editor-prosemirror/state';
import type {
  EditorAnalyticsAPI,
  RESOLVE_METHOD,
} from '@atlaskit/editor-common/analytics';
import type { Command } from '../../../types';
import { AnnotationTypes } from '@atlaskit/adf-schema';
import { createCommand } from '../pm-plugins/plugin-factory';
import { INPUT_METHOD } from '@atlaskit/editor-common/analytics';
import { isSelectionValid, getPluginState } from '../utils';
import type {
  InlineCommentAction,
  InlineCommentMap,
  InlineCommentMouseData,
} from '../pm-plugins/types';
import { ACTIONS } from '../pm-plugins/types';

import transform from './transform';
import { AnnotationSelectionType } from '../types';

export const updateInlineCommentResolvedState =
  (editorAnalyticsAPI: EditorAnalyticsAPI | undefined) =>
  (
    partialNewState: InlineCommentMap,
    resolveMethod?: RESOLVE_METHOD,
  ): Command => {
    const command: InlineCommentAction = {
      type: ACTIONS.UPDATE_INLINE_COMMENT_STATE,
      data: partialNewState,
    };

    const allResolved = Object.values(partialNewState).every((state) => state);

    if (resolveMethod && allResolved) {
      return createCommand(
        command,
        transform.addResolveAnalytics(editorAnalyticsAPI)(resolveMethod),
      );
    }

    return createCommand(command);
  };

export const closeComponent = (): Command =>
  createCommand({
    type: ACTIONS.CLOSE_COMPONENT,
  });

export const clearDirtyMark = (): Command =>
  createCommand({
    type: ACTIONS.INLINE_COMMENT_CLEAR_DIRTY_MARK,
  });

export const removeInlineCommentNearSelection =
  (id: string): Command =>
  (state, dispatch): boolean => {
    const {
      tr,
      selection: { $from },
    } = state;
    const { annotation: annotationMarkType } = state.schema.marks;

    const hasAnnotation = $from
      .marks()
      .some((mark) => mark.type === annotationMarkType);

    if (!hasAnnotation) {
      return false;
    }

    // just remove entire mark from around the node
    tr.removeMark(
      $from.start(),
      $from.end(),
      annotationMarkType.create({
        id,
        type: AnnotationTypes.INLINE_COMMENT,
      }),
    );

    if (dispatch) {
      dispatch(tr);
    }

    return true;
  };

const getDraftCommandAction: (
  drafting: boolean,
) => (state: Readonly<EditorState>) => InlineCommentAction | false = (
  drafting: boolean,
) => {
  return (editorState: EditorState) => {
    // validate selection only when entering draft mode
    if (
      drafting &&
      isSelectionValid(editorState) !== AnnotationSelectionType.VALID
    ) {
      return false;
    }

    return {
      type: ACTIONS.SET_INLINE_COMMENT_DRAFT_STATE,
      data: {
        drafting,
        editorState,
      },
    };
  };
};

export const setInlineCommentDraftState =
  (editorAnalyticsAPI: EditorAnalyticsAPI | undefined) =>
  (
    drafting: boolean,
    inputMethod:
      | INPUT_METHOD.TOOLBAR
      | INPUT_METHOD.SHORTCUT = INPUT_METHOD.TOOLBAR,
  ): Command => {
    const commandAction = getDraftCommandAction(drafting);
    return createCommand(
      commandAction,
      transform.addOpenCloseAnalytics(editorAnalyticsAPI)(
        drafting,
        inputMethod,
      ),
    );
  };

export const addInlineComment =
  (editorAnalyticsAPI: EditorAnalyticsAPI | undefined) =>
  (id: string): Command => {
    const commandAction: (editorState: EditorState) => InlineCommentAction = (
      editorState: EditorState,
    ) => ({
      type: ACTIONS.ADD_INLINE_COMMENT,
      data: {
        drafting: false,
        inlineComments: { [id]: false },
        // Auto make the newly inserted comment selected.
        // We move the selection to the head of the comment selection.
        selectedAnnotations: [{ id, type: AnnotationTypes.INLINE_COMMENT }],
        editorState,
      },
    });

    return createCommand(
      commandAction,
      transform.addInlineComment(editorAnalyticsAPI)(id),
    );
  };

export const updateMouseState = (mouseData: InlineCommentMouseData): Command =>
  createCommand({
    type: ACTIONS.INLINE_COMMENT_UPDATE_MOUSE_STATE,
    data: { mouseData },
  });

export const setSelectedAnnotation = (id: string): Command =>
  createCommand({
    type: ACTIONS.SET_SELECTED_ANNOTATION,
    data: {
      selectedAnnotations: [{ id, type: AnnotationTypes.INLINE_COMMENT }],
    },
  });

export const createAnnotation =
  (editorAnalyticsAPI: EditorAnalyticsAPI | undefined) =>
  (
    id: string,
    annotationType: AnnotationTypes = AnnotationTypes.INLINE_COMMENT,
  ): Command =>
  (state, dispatch) => {
    // don't try to add if there are is no temp highlight bookmarked
    const { bookmark } = getPluginState(state) || {};
    if (!bookmark || !dispatch) {
      return false;
    }

    if (annotationType === AnnotationTypes.INLINE_COMMENT) {
      return addInlineComment(editorAnalyticsAPI)(id)(state, dispatch);
    }

    return false;
  };

export const setInlineCommentsVisibility = (isVisible: boolean): Command => {
  return createCommand({
    type: ACTIONS.INLINE_COMMENT_SET_VISIBLE,
    data: { isVisible },
  });
};
