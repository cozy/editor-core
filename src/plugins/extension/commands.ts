import {
  removeSelectedNode,
  removeParentNodeOfType,
} from '@atlaskit/editor-prosemirror/utils';
import type { ExtensionLayout } from '@atlaskit/adf-schema';
import type { Node as PMNode } from '@atlaskit/editor-prosemirror/model';

import { createCommand } from './plugin-factory';
import type { ExtensionAction, ExtensionState } from './types';
import { getSelectedExtension } from './utils';
import { removeConnectedNodes } from '@atlaskit/editor-common/utils';
// AFP-2532 TODO: Fix automatic suppressions below
// eslint-disable-next-line @atlassian/tangerine/import/entry-points
import type {
  Parameters,
  TransformBefore,
  TransformAfter,
} from '@atlaskit/editor-common/src/extensions';
import type { ApplyChangeHandler } from '@atlaskit/editor-plugin-context-panel';

export function updateState(state: Partial<ExtensionState>) {
  return createCommand({
    type: 'UPDATE_STATE',
    data: state,
  });
}

export function setEditingContextToContextPanel<
  T extends Parameters = Parameters,
>(
  processParametersBefore: TransformBefore<T>,
  processParametersAfter: TransformAfter<T>,
  applyChangeToContextPanel: ApplyChangeHandler | undefined,
) {
  return createCommand<ExtensionAction<T>>(
    {
      type: 'UPDATE_STATE',
      data: {
        showContextPanel: true,
        processParametersBefore,
        processParametersAfter,
      },
    },
    applyChangeToContextPanel,
  );
}

export const clearEditingContext = (
  applyChangeToContextPanel: ApplyChangeHandler | undefined,
) =>
  createCommand(
    {
      type: 'UPDATE_STATE',
      data: {
        showContextPanel: false,
        processParametersBefore: undefined,
        processParametersAfter: undefined,
      },
    },
    applyChangeToContextPanel,
  );

export const forceAutoSave =
  (applyChangeToContextPanel: ApplyChangeHandler | undefined) =>
  (resolve: () => void, reject?: (reason?: any) => void) =>
    createCommand(
      {
        type: 'UPDATE_STATE',
        data: { autoSaveResolve: resolve, autoSaveReject: reject },
      },
      applyChangeToContextPanel,
    );

export const updateExtensionLayout = (layout: ExtensionLayout) =>
  createCommand({ type: 'UPDATE_STATE', data: { layout } }, (tr, state) => {
    const selectedExtension = getSelectedExtension(state, true);

    if (selectedExtension) {
      const trWithNewNodeMarkup = tr.setNodeMarkup(
        selectedExtension.pos,
        undefined,
        {
          ...selectedExtension.node.attrs,
          layout,
        },
      );
      trWithNewNodeMarkup.setMeta('scrollIntoView', false);
      return trWithNewNodeMarkup;
    }

    return tr;
  });

export const removeExtension = () =>
  createCommand(
    {
      type: 'UPDATE_STATE',
      data: { element: undefined },
    },
    (tr, state) => {
      if (getSelectedExtension(state)) {
        return removeSelectedNode(tr);
      } else {
        return removeParentNodeOfType(state.schema.nodes.bodiedExtension)(tr);
      }
    },
  );

export const removeDescendantNodes = (sourceNode?: PMNode) =>
  createCommand(
    {
      type: 'UPDATE_STATE',
      data: { element: undefined },
    },
    (tr, state) => {
      return sourceNode ? removeConnectedNodes(state, sourceNode) : tr;
    },
  );
