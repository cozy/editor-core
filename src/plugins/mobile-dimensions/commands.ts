import { MobileDimensionsActionTypes } from './actions';
import { createCommand } from './plugin-factory';

export const setWindowHeight = (windowHeight: number) =>
  createCommand({
    type: MobileDimensionsActionTypes.SET_WINDOW_HEIGHT,
    windowHeight,
  });

export const setMobilePaddingTop = (paddingTop: number) =>
  createCommand({
    type: MobileDimensionsActionTypes.SET_MOBILE_PADDING_TOP,
    paddingTop,
  });

export const setIsExpanded = (isExpanded: boolean) =>
  createCommand({
    type: MobileDimensionsActionTypes.SET_IS_EXPANDED,
    isExpanded,
  });
