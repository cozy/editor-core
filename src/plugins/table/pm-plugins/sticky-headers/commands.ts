import { createCommand, RowStickyState } from './plugin-state';

export const updateStickyState = (rowState: RowStickyState) =>
  createCommand({ name: 'UPDATE', state: rowState }, tr =>
    tr.setMeta('addToHistory', false),
  );

export const removeStickyState = (pos: number) =>
  createCommand({ name: 'REMOVE', pos }, tr =>
    tr.setMeta('addToHistory', false),
  );
