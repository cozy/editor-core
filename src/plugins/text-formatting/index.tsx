import React from 'react';

import {
  code,
  em,
  strike,
  strong,
  subsup,
  underline,
} from '@atlaskit/adf-schema';

import { EditorPlugin } from '../../types';
import { ButtonGroup } from '../../ui/styles';
import WithPluginState from '../../ui/WithPluginState';

import {
  plugin as clearFormattingPlugin,
  pluginKey as clearFormattingPluginKey,
} from './pm-plugins/clear-formatting';
import clearFormattingKeymapPlugin from './pm-plugins/clear-formatting-keymap';
import textFormattingCursorPlugin from './pm-plugins/cursor';
import textFormattingInputRulePlugin from './pm-plugins/input-rule';
import keymapPlugin from './pm-plugins/keymap';
import {
  plugin as textFormattingPlugin,
  pluginKey as textFormattingPluginKey,
} from './pm-plugins/main';
import textFormattingSmartInputRulePlugin from './pm-plugins/smart-input-rule';
import { TextFormattingOptions } from './types';
import ToolbarAdvancedTextFormatting from './ui/ToolbarAdvancedTextFormatting';
import ToolbarTextFormatting from './ui/ToolbarTextFormatting';
import { akEditorMobileMaxWidth } from '@atlaskit/editor-shared-styles';
import { pluginKey as widthPluginKey } from '../../plugins/width/index';

const textFormatting = (options: TextFormattingOptions = {}): EditorPlugin => ({
  name: 'textFormatting',

  marks() {
    return [
      { name: 'em', mark: em },
      { name: 'strong', mark: strong },
      { name: 'strike', mark: strike },
    ]
      .concat(options.disableCode ? [] : { name: 'code', mark: code })
      .concat(
        options.disableSuperscriptAndSubscript
          ? []
          : { name: 'subsup', mark: subsup },
      )
      .concat(
        options.disableUnderline ? [] : { name: 'underline', mark: underline },
      );
  },

  pmPlugins() {
    return [
      {
        name: 'textFormatting',
        plugin: ({ dispatch }) => textFormattingPlugin(dispatch),
      },
      {
        name: 'textFormattingCursor',
        plugin: () => textFormattingCursorPlugin,
      },
      {
        name: 'textFormattingInputRule',
        plugin: ({ schema }) => textFormattingInputRulePlugin(schema),
      },
      {
        name: 'textFormattingSmartRule',
        plugin: () =>
          !options.disableSmartTextCompletion
            ? textFormattingSmartInputRulePlugin
            : undefined,
      },
      {
        name: 'textFormattingClear',
        plugin: ({ dispatch }) => clearFormattingPlugin(dispatch),
      },
      {
        name: 'textFormattingClearKeymap',
        plugin: () => clearFormattingKeymapPlugin(),
      },
      {
        name: 'textFormattingKeymap',
        plugin: ({ schema }) => keymapPlugin(schema),
      },
    ];
  },

  primaryToolbarComponent({
    editorView,
    popupsMountPoint,
    popupsScrollableElement,
    isToolbarReducedSpacing,
    disabled,
  }) {
    return (
      <WithPluginState
        plugins={{
          textFormattingState: textFormattingPluginKey,
          clearFormattingState: clearFormattingPluginKey,
          widthState: widthPluginKey,
        }}
        render={({
          textFormattingState,
          clearFormattingState,
          widthState,
        }): any => {
          return (
            <ButtonGroup width={isToolbarReducedSpacing ? 'small' : 'large'}>
              {/* render toolbar buttons for non-mobile views */}
              {widthState.width > akEditorMobileMaxWidth && (
                <ToolbarTextFormatting
                  disabled={disabled}
                  editorView={editorView}
                  textFormattingState={textFormattingState}
                  isReducedSpacing={isToolbarReducedSpacing}
                />
              )}
              <ToolbarAdvancedTextFormatting
                editorView={editorView}
                isDisabled={disabled}
                isReducedSpacing={isToolbarReducedSpacing}
                textFormattingState={textFormattingState}
                clearFormattingState={clearFormattingState}
                popupsMountPoint={popupsMountPoint}
                popupsScrollableElement={popupsScrollableElement}
              />
            </ButtonGroup>
          );
        }}
      />
    );
  },
});

export default textFormatting;
