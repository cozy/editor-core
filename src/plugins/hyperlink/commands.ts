import { Command } from '../../types';
import { normalizeUrl } from './utils';
import { stateKey, LinkAction } from './pm-plugins/main';
import { EditorState, Selection } from 'prosemirror-state';
import { filter, Predicate } from '../../utils/commands';
import { Mark, Node, ResolvedPos } from 'prosemirror-model';
import {
  addAnalytics,
  ACTION,
  ACTION_SUBJECT,
  INPUT_METHOD,
  EVENT_TYPE,
  ACTION_SUBJECT_ID,
  withAnalytics,
} from '../analytics';
import { queueCardsFromChangedTr } from '../card/pm-plugins/doc';
import { LinkInputType } from './types';
import { getLinkCreationAnalyticsEvent } from './analytics';
import { LinkAttributes } from '@atlaskit/adf-schema';

export function isTextAtPos(pos: number): Predicate {
  return (state: EditorState) => {
    const node = state.doc.nodeAt(pos);
    return !!node && node.isText;
  };
}

export function isLinkAtPos(pos: number): Predicate {
  return (state: EditorState) => {
    const node = state.doc.nodeAt(pos);
    return !!node && state.schema.marks.link.isInSet(node.marks);
  };
}

export function setLinkHref(
  href: string,
  pos: number,
  to?: number,
  isTabPressed?: boolean,
): Command {
  return filter(isTextAtPos(pos), (state, dispatch) => {
    const $pos = state.doc.resolve(pos);
    const node = state.doc.nodeAt(pos) as Node;
    const linkMark = state.schema.marks.link;
    const mark = linkMark.isInSet(node.marks) as Mark | undefined;
    const url = normalizeUrl(href);

    if (mark && mark.attrs.href === url) {
      return false;
    }

    const rightBound =
      to && pos !== to ? to : pos - $pos.textOffset + node.nodeSize;

    const tr = state.tr.removeMark(pos, rightBound, linkMark);

    if (href.trim()) {
      tr.addMark(
        pos,
        rightBound,
        linkMark.create({
          ...((mark && mark.attrs) || {}),
          href: url,
        }),
      );
    }
    if (!isTabPressed) {
      tr.setMeta(stateKey, { type: LinkAction.HIDE_TOOLBAR });
    }

    if (dispatch) {
      dispatch(tr);
    }
    return true;
  });
}

export function updateLink(
  href: string,
  text: string,
  pos: number,
  to?: number,
): Command {
  return (state, dispatch) => {
    const $pos: ResolvedPos = state.doc.resolve(pos);
    const node: Node | null | undefined = state.doc.nodeAt(pos);
    if (!node) {
      return false;
    }
    const url = normalizeUrl(href);

    if (!url) {
      return false;
    }

    const mark: Mark = state.schema.marks.link.isInSet(node.marks);
    const linkMark = state.schema.marks.link;

    const rightBound =
      to && pos !== to ? to : pos - $pos.textOffset + node.nodeSize;
    const tr = state.tr;
    tr.insertText(text, pos, rightBound);
    // Casting to LinkAttributes to prevent wrong attributes been passed (Example ED-7951)
    const linkAttrs: LinkAttributes = {
      ...((mark && (mark.attrs as LinkAttributes)) || {}),
      href: url,
    };
    tr.addMark(pos, pos + text.length, linkMark.create(linkAttrs));
    tr.setMeta(stateKey, { type: LinkAction.HIDE_TOOLBAR });
    if (dispatch) {
      dispatch(tr);
    }
    return true;
  };
}

export function setLinkText(text: string, pos: number, to?: number): Command {
  return filter(isLinkAtPos(pos), (state, dispatch) => {
    const $pos = state.doc.resolve(pos);
    const node = state.doc.nodeAt(pos) as Node;
    const mark = state.schema.marks.link.isInSet(node.marks) as Mark;
    if (node && text.length > 0 && text !== node.text) {
      const rightBound =
        to && pos !== to ? to : pos - $pos.textOffset + node.nodeSize;
      const tr = state.tr;

      tr.insertText(text, pos, rightBound);
      tr.addMark(pos, pos + text.length, mark);
      tr.setMeta(stateKey, { type: LinkAction.HIDE_TOOLBAR });

      if (dispatch) {
        dispatch(tr);
      }
      return true;
    }
    return false;
  });
}

export function insertLink(
  from: number,
  to: number,
  incomingHref: string,
  incomingTitle?: string,
  displayText?: string,
  source?: LinkInputType,
): Command {
  return (state, dispatch) => {
    const link = state.schema.marks.link;
    if (incomingHref.trim()) {
      const { tr } = state;
      const normalizedUrl = normalizeUrl(incomingHref);
      // NB: in this context, `currentText` represents text which has been
      // highlighted in the Editor, upon which a link is is being added.
      const currentText = stateKey.getState(state).activeText;

      let markEnd = to;
      const text = displayText || incomingTitle || incomingHref;
      if (!displayText || displayText !== currentText) {
        tr.insertText(text, from, to);
        markEnd = from + text.length;
      }

      tr.addMark(from, markEnd, link.create({ href: normalizedUrl }));
      tr.setSelection(Selection.near(tr.doc.resolve(markEnd)));

      //Create a smart link when the user doesn't provide a title
      if (isSmartLink(incomingHref, source, incomingTitle, displayText)) {
        queueCardsFromChangedTr(state, tr, source!, false);
      }

      tr.setMeta(stateKey, { type: LinkAction.HIDE_TOOLBAR });
      if (dispatch) {
        dispatch(tr);
      }
      return true;
    }
    return false;
  };
}

export const insertLinkWithAnalytics = (
  inputMethod: LinkInputType,
  from: number,
  to: number,
  href: string,
  title?: string | undefined,
  displayText?: string | undefined,
) => {
  // If we are inserting a smart link, skip insert hyperlink analytics
  if (isSmartLink(href, inputMethod, title, displayText)) {
    return insertLink(from, to, href, title, displayText, inputMethod);
  } else {
    return withAnalytics(getLinkCreationAnalyticsEvent(inputMethod, href))(
      insertLink(from, to, href, title, displayText, inputMethod),
    );
  }
};

const isSmartLink = (
  href: string,
  inputMethod?: LinkInputType,
  title?: string,
  displayText?: string,
) =>
  // NB: A link is a Smart Link if:
  // - it is inserted from the CMD + K menu without display text
  // - it is inserted manually (by typing) without display text and a title
  // - it is inserted using either approach and the display text, href are the same
  (inputMethod && inputMethod === INPUT_METHOD.TYPEAHEAD && !displayText) ||
  (inputMethod &&
    inputMethod === INPUT_METHOD.MANUAL &&
    !title &&
    !displayText) ||
  displayText === href;

export function removeLink(pos: number): Command {
  return setLinkHref('', pos);
}

export function editInsertedLink(): Command {
  return (state, dispatch) => {
    if (dispatch) {
      dispatch(
        state.tr.setMeta(stateKey, {
          type: LinkAction.EDIT_INSERTED_TOOLBAR,
          inputMethod: INPUT_METHOD.FLOATING_TB,
        }),
      );
    }
    return true;
  };
}

export function showLinkToolbar(
  inputMethod:
    | INPUT_METHOD.TOOLBAR
    | INPUT_METHOD.QUICK_INSERT
    | INPUT_METHOD.SHORTCUT
    | INPUT_METHOD.INSERT_MENU = INPUT_METHOD.TOOLBAR,
): Command {
  return function (state, dispatch) {
    if (dispatch) {
      let tr = state.tr.setMeta(stateKey, {
        type: LinkAction.SHOW_INSERT_TOOLBAR,
        inputMethod,
      });
      tr = addAnalytics(state, tr, {
        action: ACTION.INVOKED,
        actionSubject: ACTION_SUBJECT.TYPEAHEAD,
        actionSubjectId: ACTION_SUBJECT_ID.TYPEAHEAD_LINK,
        attributes: { inputMethod },
        eventType: EVENT_TYPE.UI,
      });
      dispatch(tr);
    }
    return true;
  };
}

export function hideLinkToolbar(): Command {
  return function (state, dispatch) {
    if (dispatch) {
      dispatch(state.tr.setMeta(stateKey, { type: LinkAction.HIDE_TOOLBAR }));
    }
    return true;
  };
}
