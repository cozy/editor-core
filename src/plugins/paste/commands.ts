import { createCommand } from './pm-plugins/plugin-factory';
import { PastePluginActionTypes as ActionTypes } from './actions';
import type {
  Node,
  NodeType,
  Schema,
} from '@atlaskit/editor-prosemirror/model';
import { Fragment, Slice } from '@atlaskit/editor-prosemirror/model';
import {
  isListNode,
  mapChildren,
  mapSlice,
} from '@atlaskit/editor-common/utils';
import type { Transaction } from '@atlaskit/editor-prosemirror/state';
import { EditorState } from '@atlaskit/editor-prosemirror/state';
import { autoJoin } from '@atlaskit/editor-prosemirror/commands';
import type { Mark } from '@atlaskit/editor-prosemirror/model';

/**
 * Use this to register macro link positions during a paste operation, that you
 * want to track in a document over time, through any document changes.
 *
 * @param positions a map of string keys (custom position references) and position values e.g. { ['my-key-1']: 11 }
 *
 * **Context**: This is neccessary if there is an async process or an unknown period of time
 * between obtaining an original position, and wanting to know about what its final eventual
 * value. In that scenario, positions will need to be actively tracked and mapped in plugin
 * state so that they can be mapped through any other independent document change transactions being
 * dispatched to the editor that could affect their value.
 */
export const startTrackingPastedMacroPositions = (pastedMacroPositions: {
  [key: string]: number;
}) =>
  createCommand(() => {
    return {
      type: ActionTypes.START_TRACKING_PASTED_MACRO_POSITIONS,
      pastedMacroPositions,
    };
  });

export const stopTrackingPastedMacroPositions = (
  pastedMacroPositionKeys: string[],
) =>
  createCommand(() => {
    return {
      type: ActionTypes.STOP_TRACKING_PASTED_MACRO_POSITIONS,
      pastedMacroPositionKeys,
    };
  });

// matchers for text lists
const bullets = /^\s*[\*\-\u2022](\s+|\s+$)/;
const numbers = /^\s*\d[\.\)](\s+|$)/;

const getListType = (node: Node, schema: Schema): [NodeType, number] | null => {
  if (!node.text) {
    return null;
  }

  const { bulletList, orderedList } = schema.nodes;

  return [
    {
      node: bulletList,
      matcher: bullets,
    },
    {
      node: orderedList,
      matcher: numbers,
    },
  ].reduce((lastMatch: [NodeType, number] | null, listType) => {
    if (lastMatch) {
      return lastMatch;
    }

    const match = node.text!.match(listType.matcher);
    return match ? [listType.node, match[0].length] : lastMatch;
  }, null);
};

const extractListFromParagraph = (
  node: Node,
  parent: Node | null,
  schema: Schema,
): Fragment => {
  const { hardBreak, bulletList, orderedList } = schema.nodes;
  const content: Array<Node> = mapChildren(node.content, (node) => node);

  const listTypes = [bulletList, orderedList];

  // wrap each line into a listItem and a containing list
  const listified = content
    .map((child, index) => {
      const listMatch = getListType(child, schema);
      const prevChild = index > 0 && content[index - 1];

      // only extract list when preceded by a hardbreak
      if (prevChild && prevChild.type !== hardBreak) {
        return child;
      }

      if (!listMatch || !child.text) {
        return child;
      }

      const [nodeType, length] = listMatch;

      // convert to list item
      const newText = child.text.substr(length);
      const listItemNode = schema.nodes.listItem.createAndFill(
        undefined,
        schema.nodes.paragraph.createChecked(
          undefined,
          newText.length ? schema.text(newText) : undefined,
        ),
      );

      if (!listItemNode) {
        return child;
      }

      const newList = nodeType.createChecked(undefined, [listItemNode]);
      // Check whether our new list is valid content in our current structure,
      // otherwise dont convert.
      if (parent && !parent.type.validContent(Fragment.from(newList))) {
        return child;
      }

      return newList;
    })
    .filter((child, idx, arr) => {
      // remove hardBreaks that have a list node on either side

      // wasn't hardBreak, leave as-is
      if (child.type !== hardBreak) {
        return child;
      }

      if (idx > 0 && listTypes.indexOf(arr[idx - 1].type) > -1) {
        // list node on the left
        return null;
      }

      if (idx < arr.length - 1 && listTypes.indexOf(arr[idx + 1].type) > -1) {
        // list node on the right
        return null;
      }

      return child;
    });

  // try to join
  const mockState = EditorState.create({
    schema,
  });

  let joinedListsTr: Transaction | undefined;
  const mockDispatch = (tr: Transaction) => {
    joinedListsTr = tr;
  };

  autoJoin(
    (state, dispatch) => {
      if (!dispatch) {
        return false;
      }

      // Return false to prevent replaceWith from wrapping the text node in a paragraph
      // paragraph since that will be done later. If it's done here, it will fail
      // the paragraph.validContent check.
      // Dont return false if there are lists, as they arent validContent for paragraphs
      // and will result in hanging textNodes
      const containsList = listified.some(
        (node) => node.type === bulletList || node.type === orderedList,
      );
      if (listified.some((node) => node.isText) && !containsList) {
        return false;
      }

      dispatch(state.tr.replaceWith(0, 2, listified));
      return true;
    },
    (before, after) => isListNode(before) && isListNode(after),
  )(mockState, mockDispatch);

  const fragment = joinedListsTr
    ? joinedListsTr.doc.content
    : Fragment.from(listified);

  // try to re-wrap fragment in paragraph (which is the original node we unwrapped)
  const { paragraph } = schema.nodes;
  if (paragraph.validContent(fragment)) {
    return Fragment.from(paragraph.create(node.attrs, fragment, node.marks));
  }

  // fragment now contains other nodes, get Prosemirror to wrap with ContentMatch later
  return fragment;
};

// above will wrap everything in paragraphs for us
export const upgradeTextToLists = (slice: Slice, schema: Schema): Slice => {
  return mapSlice(slice, (node, parent) => {
    if (node.type === schema.nodes.paragraph) {
      return extractListFromParagraph(node, parent, schema);
    }

    return node;
  });
};

export const splitParagraphs = (slice: Slice, schema: Schema): Slice => {
  // exclude Text nodes with a code mark, since we transform those later
  // into a codeblock
  let hasCodeMark = false;
  slice.content.forEach((child) => {
    hasCodeMark =
      hasCodeMark ||
      child.marks.some((mark) => mark.type === schema.marks.code);
  });

  // slice might just be a raw text string
  if (schema.nodes.paragraph.validContent(slice.content) && !hasCodeMark) {
    const replSlice = splitIntoParagraphs({ fragment: slice.content, schema });
    return new Slice(replSlice, slice.openStart + 1, slice.openEnd + 1);
  }

  return mapSlice(slice, (node) => {
    if (node.type === schema.nodes.paragraph) {
      return splitIntoParagraphs({
        fragment: node.content,
        blockMarks: node.marks,
        schema,
      });
    }

    return node;
  });
};

/**
 * Walks the slice, creating paragraphs that were previously separated by hardbreaks.
 * Returns the original paragraph node (as a fragment), or a fragment containing multiple nodes.
 */
export const splitIntoParagraphs = ({
  fragment,
  blockMarks = [],
  schema,
}: {
  fragment: Fragment;
  blockMarks?: readonly Mark[];
  schema: Schema;
}): Fragment => {
  const paragraphs = [];
  let curChildren: Array<Node> = [];
  let lastNode: Node | null = null;

  const { hardBreak, paragraph } = schema.nodes;

  fragment.forEach((node, i) => {
    const isNodeValidContentForParagraph = schema.nodes.paragraph.validContent(
      Fragment.from(node),
    );

    if (!isNodeValidContentForParagraph) {
      paragraphs.push(node);
      return;
    }
    // ED-14725 Fixed the issue that it make duplicated line
    // when pasting <br /> from google docs.
    if (i === 0 && node.type === hardBreak) {
      paragraphs.push(
        paragraph.createChecked(undefined, curChildren, [...blockMarks]),
      );
      lastNode = node;
      return;
    } else if (
      lastNode &&
      lastNode.type === hardBreak &&
      node.type === hardBreak
    ) {
      // double hardbreak

      // backtrack a little; remove the trailing hardbreak we added last loop
      curChildren.pop();

      // create a new paragraph
      paragraphs.push(
        paragraph.createChecked(undefined, curChildren, [...blockMarks]),
      );
      curChildren = [];
      return;
    }

    // add to this paragraph
    curChildren.push(node);
    lastNode = node;
  });

  if (curChildren.length) {
    paragraphs.push(
      paragraph.createChecked(undefined, curChildren, [...blockMarks]),
    );
  }

  return Fragment.from(
    paragraphs.length
      ? paragraphs
      : [paragraph.createAndFill(undefined, undefined, [...blockMarks])!],
  );
};
