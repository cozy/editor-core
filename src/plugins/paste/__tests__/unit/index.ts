import {
  Preset,
  LightEditorPlugin,
  createProsemirrorEditorFactory,
} from '@atlaskit/editor-test-helpers/create-prosemirror-editor';
import dispatchPasteEvent from '@atlaskit/editor-test-helpers/dispatch-paste-event';
import { doc, p } from '@atlaskit/editor-test-helpers/schema-builder';
import { EditorView } from 'prosemirror-view';
import { toggleStrong } from '../../../text-formatting/commands/text-formatting';
import pastePlugin from '../../index';
import blockTypePlugin from '../../../block-type';
import { textFormattingPlugin } from '../../../index';

describe('Paste plugin', () => {
  const createEditor = createProsemirrorEditorFactory();
  let editorView: EditorView;

  const editor = (doc: any) =>
    createEditor({
      doc,
      preset: new Preset<LightEditorPlugin>()
        .add([pastePlugin, {}])
        .add(blockTypePlugin)
        .add(textFormattingPlugin),
    });

  beforeEach(() => {
    ({ editorView } = editor(doc(p('{<>}'))));
  });

  describe('cmd+shift+v', () => {
    const paste = () => {
      dispatchPasteEvent(
        editorView,
        {
          html: `<meta charset='utf-8'><div style="color: #333333;background-color: #f5f5f5;font-family: Menlo, Monaco, 'Courier New', monospace;font-weight: normal;font-size: 12px;line-height: 18px;white-space: pre;"><div><span style="color: #4b69c6;">export</span><span style="color: #333333;"> </span><span style="color: #7a3e9d;">interface</span><span style="color: #333333;"> </span><span style="color: #7a3e9d;font-weight: bold;">PasteContent</span></div></div>`,
          plain: 'export interface PasteContent',
        },
        { shift: true },
      );
    };

    it('pastes plain text', () => {
      paste();
      expect(editorView.state.doc).toMatchSnapshot();
    });

    it('preserves current formatting when pasting', () => {
      toggleStrong()(editorView.state, editorView.dispatch);
      paste();
      expect(editorView.state.doc).toMatchSnapshot();
    });
  });
});
