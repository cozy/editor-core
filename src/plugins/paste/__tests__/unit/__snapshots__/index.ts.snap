// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Paste plugin cmd+shift+v pastes plain text 1`] = `
Object {
  "content": Array [
    Object {
      "content": Array [
        Object {
          "text": "export interface PasteContent",
          "type": "text",
        },
      ],
      "type": "paragraph",
    },
  ],
  "type": "doc",
}
`;

exports[`Paste plugin cmd+shift+v preserves current formatting when pasting 1`] = `
Object {
  "content": Array [
    Object {
      "content": Array [
        Object {
          "marks": Array [
            Object {
              "type": "strong",
            },
          ],
          "text": "export interface PasteContent",
          "type": "text",
        },
      ],
      "type": "paragraph",
    },
  ],
  "type": "doc",
}
`;
